### PR TITLE
Add incident management worklow

### PR DESCRIPTION
**Pull request by louisfelix**
This PR adds incident management workflow. It is described in the README.

Internal incident modelling has changed because a simple incident `struct` did not met the requirements anymore. ServiceNow incident model returned by REST API is not always consistent depending if you POST or GET an incident. But even more important, there is a need for dynamic (configurable) incident fields; `incident_group_key_field` is the first one, but we can easily imagine in the future that the webhook will support custom incident fields that a user may have configured in his ServiceNow installation.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -117,32 +141,84 @@ func loadConfig(configFile string) Config {
...
 	if err != nil {
-		log.Errorf("Error while creating incident: %v", err)
 		return err
 	}
 
+	var incident Incident
+	if len(incidents) > 0 {
+		incident = incidents[0]
+	}
+	if len(incidents) > 1 {
```
Comment (vincentcreusotfx) [inline main.go]: You should put this condition inside the other

File: main.go [hunk 2]
```diff
@@ -117,32 +141,84 @@ func loadConfig(configFile string) Config {
...
+		log.Infof("Found existing incident (%s), with state %s, for resolved alert group key: %s", incident.GetNumber(), incident.GetState(), getGroupKey(data))
+		if _, err := serviceNow.UpdateIncident(incidentParam, incident.GetSysID()); err != nil {
+			return err
+		}
+	}
 	return nil
 }
 
-func dataToIncident(data template.Data) Incident {
+func dataToIncidentParam(data template.Data) IncidentParam {
```
Comment (vincentcreusotfx) [inline main.go]: name of the func is too vague. I understand it's a data object to something else but you shoud put something more precise than data

File: main.go [hunk 3]
```diff
@@ -183,7 +259,22 @@ func dataToIncident(data template.Data) Incident {
...
-	return incident
+
+	return incidentParam
+}
+
+func getGroupKey(data template.Data) string {
+	return fmt.Sprintf("%v", data.GroupLabels.SortedPairs())
+}
+
+func jsonNumberInSlice(n json.Number, list []json.Number) bool {
```
Comment (vincentcreusotfx) [inline main.go]: name of the func should tell what it does I'd say isJsonNumberInSlice or maybe isNumberInSlice would be enough

File: main.go (hunk 1 above)
Comment (louisfelix) [reply main.go]: I agree. Done.

File: main.go (hunk 2 above)
Comment (louisfelix) [reply main.go]: Yeah ok. Done.

File: main.go (hunk 3 above)
Comment (louisfelix) [reply main.go]: Yeah ok. Done.

Comment (vincentcreusotfx) [reply main.go]: Would have put maybe contailsNumber(json.Number, []json.Number)
But thinking of it there is another way to do it. I'm not sure if you're using this method several times but if it's the case, you should create a map and do a _,contains=map[key] which would be more efficient

File: main.go [hunk 4]
```diff
@@ -59,8 +83,8 @@ func main() {
 	kingpin.HelpFlag.Short('h')
 	kingpin.Parse()
 
-	config = loadConfig(*configFile)
-	createSnClient(config)
+	loadConfig(*configFile)
```
Comment (vincentcreusotfx) [inline main.go]: Wouldn't it be better to return an error and centralize the error management in the main method ?

File: main.go [hunk 5]
```diff
@@ -174,7 +251,7 @@ func dataToIncident(data template.Data) Incident {
 		commentBuilder.WriteString(fmt.Sprintf("\n\n%s", alertBuilder.String()))
 	}
 
-	incident := Incident{
+	incidentParam := IncidentParam{
 		AssignmentGroup:  config.DefaultIncident.AssignmentGroup,
```
Comment (vincentcreusotfx) [inline main.go]: What if DefaultIncident is nil ?

File: main.go (hunk 3 above)
Comment (louisfelix) [reply main.go]: Well you suggested isJsonNumberInSlice, then isNumberInSlice and now contailsNumber ;-)
Well I changed to a map, even though performance gain is almost nil here, it is somewhat cleaner. Done.

File: main.go (hunk 5 above)
Comment (louisfelix) [reply main.go]: It will not, it would be loaded as zero value struct if config is missing.

File: main.go (hunk 3 above)
Comment (vincentcreusotfx) [reply main.go]: Sorry ... thanks

File: main.go (hunk 5 above)
Comment (vincentcreusotfx) [reply main.go]: Ok then

File: servicenow.go [hunk 6]
```diff
@@ -17,57 +17,95 @@ const (
...
+type IncidentResponse map[string]interface{}
+
+// GetResult returns the incident from the IncidentResponse
+func (ir IncidentResponse) GetResult() Incident {
+	var incident Incident = ir["result"].(map[string]interface{})
+	return incident
+}
+
+// IncidentsResponse is a model of an API response contaning multiple incidents
+type IncidentsResponse map[string]interface{}
```
Comment (vincentcreusotfx) [inline servicenow.go]: So if I understand well, the return of the HTTP call is just a map of keys assigned to something, then you just unmarshall the whole response to have all the values and keys like dynamically ?

Comment (louisfelix) [reply servicenow.go]: Yes. This is one of the documented way to achieve highly dynamic yaml parsing.

Comment (vincentcreusotfx) [reply servicenow.go]: Nice

File: main.go (hunk 4 above)
Comment (louisfelix) [reply main.go]: Even though this was not a change made in the PR, I will take a look.

Comment (louisfelix) [reply main.go]: Done.

**Pull request merged**